securedrop/config_tool.py

```diff
@@ -0,0 +1,327 @@
+#!/usr/bin/env python
+
+import argparse
+import logging
+import re
+import sys
+import yaml
+
+from os import path
+from yaml.scanner import ScannerError
+
+
+"""
+A CLI tool to help admins configure their SecureDrop instances. Uses argparse, so run this file from
+your terminal with --help to get instructions on how to use this.
+"""
+
+
+def init_logger():
+    l = logging.getLogger(__file__)
+    l.setLevel(logging.DEBUG)
+    ch = logging.StreamHandler(sys.stdout)
+    ch.setLevel(logging.DEBUG)
+    ch.setFormatter(logging.Formatter('%(levelname)s - %(message)s'))
+    l.addHandler(ch)
+    return l
+
+log = init_logger()
+
+
+class ConfigToolException(Exception):
+
+    def __init__(self, msg=None):
+        super(ConfigToolException, self).__init__(msg)
+
+
+def check_ip_address_valid(ip):  # "good enough"
+    m = re.match(r"^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$", ip)
+    if bool(m) and all(map(lambda n: 0 <= int(n) <= 255, m.groups())):
+        log.debug("'{ip}' was a valid ipv4 address".format(ip=ip))
+        return True
+    else:
+        log.error("'{ip}' was not a valid ipv4 address".format(ip=ip))
+        return False
+
+
+def check_ssh_users(users):
+    def check_user(user):
+        regex = r'^[a-z][-a-z0-9]*$'
+        if isinstance(user, str) and re.match(regex, user):
+            log.debug("'{user}' was a valid Unix user name".format(user=user))
+            return True
+        else:
+            log.debug("'{user}' was not valid Unix user name. Regex: {regex}".format(user=user, regex=regex))
+            return False
+
+    if isinstance(users, str):
+        return check_user(users)
+    elif isinstance(users, list):
+        return all(map(check_user, users))
+    else:
+        log.error('SSH users need to be a string or list of strings')
+        return False
+
+
+def check_dns_servers(servers):
+    if isinstance(servers, str):
+        return check_ip_address_valid(servers)
+    elif isinstance(servers, list):
+        return all(map(check_ip_address_valid, servers))
+    else:
+        log.error('DNS servers need to be a string or list of strings')
+        return False
+
+
+def check_gpg_public_key(key_str):
+    key_lines = key_str.splitlines()
+    return len(key_lines) >= 2 and \
+        key_lines[0] == '-----BEGIN PGP PUBLIC KEY BLOCK-----' in key_str and \
+        (key_lines[-1] == '-----END PGP PUBLIC KEY BLOCK-----' or
+         key_lines[-2] == '-----END PGP PUBLIC KEY BLOCK-----')
+
+
+def check_gpg_public_key_fingerprint(fpr):
+    return bool(re.match(r'^[0-9A-F]{40}$', fpr))
+
+
+def check_email_address(addr):
+    return '@' in addr
```

Comment:
While not suggesting you do this 100% correct (if such a thing is possible in this case :smirk_cat:), it's probably worth it to be a little more strict. I think @jesusaurus has the right idea here that would help catch, e.g., a `,` typed instead of a `.`.